ddl/index.go

```diff
@@ -700,6 +685,431 @@ func (w *worker) doBackfillIndexTaskInTxn(t table.Table, txn kv.Transaction, col
+		func(handle int64, recordKey kv.Key, rawRow []byte) (bool, error) {
+			// Don't create endHandle index.
+			handleOutOfRange = handle >= taskRange.endHandle
+			if handleOutOfRange || len(w.idxRecords) >= w.batchCnt {
+				return false, nil
+			}
+
+			idxRecord, err1 := w.getIndexRecord(handle, recordKey, rawRow)
+			if err1 != nil {
+				return false, errors.Trace(err1)
+			}
+
+			w.idxRecords = append(w.idxRecords, idxRecord)
+			return true, nil
+		})
+
+	log.Debugf("[ddl] txn %v fetches handle info %v, takes time %v", txn.StartTS(), taskRange, time.Since(startTime))
+	return w.idxRecords, handleOutOfRange, errors.Trace(err)
+}
+
+// backfillIndexInTxn will backfill table index in a transaction, lock corresponding rowKey, if the value of rowKey is changed,
+// indicate that index columns values may changed, index is not allowed to be added, so the txn will rollback and retry.
+// backfillIndexInTxn will add w.batchCnt indices once, default value of w.batchCnt is 128.
+// TODO: make w.batchCnt can be modified by system variable.
+func (w *addIndexWorker) backfillIndexInTxn(handleRange reorgIndexTask) (nextHandle int64, addedCount, scanCount int, errInTxn error) {
+	addedCount = 0
+	scanCount = 0
+	errInTxn = kv.RunInNewTxn(w.sessCtx.GetStore(), true, func(txn kv.Transaction) error {
+		txn.SetOption(kv.Priority, kv.PriorityLow)
+		idxRecords, handleOutOfRange, err := w.fetchRowColVals(txn, handleRange)
+
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for _, idxRecord := range idxRecords {
+			err := txn.LockKeys(idxRecord.key)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			scanCount++
+
+			// Create the index.
+			// TODO: backfill unique-key will check constraint every row, we can speed up this case by using batch check.
+			handle, err := w.index.Create(w.sessCtx, txn, idxRecord.vals, idxRecord.handle)
+			if err != nil {
+				if kv.ErrKeyExists.Equal(err) && idxRecord.handle == handle {
+					// Index already exists, skip it.
+					continue
+				}
+
+				return errors.Trace(err)
+			}
+			addedCount++
+		}
+
+		if handleOutOfRange || len(idxRecords) == 0 {
+			nextHandle = handleRange.endHandle
+		} else {
+			nextHandle = idxRecords[len(idxRecords)-1].handle + 1
+		}
+		return nil
+	})
+
+	return
+}
+
+// handleBackfillTask backfill range [task.startHandle, task.endHandle) handle's index to table.
+func (w *addIndexWorker) handleBackfillTask(task *reorgIndexTask) *addIndexResult {
+	handleRange := *task
+	result := &addIndexResult{addedCount: 0, nextHandle: handleRange.startHandle, err: nil}
+	lastLogCount := 0
+	for {
+		addedCount := 0
+		nextHandle, addedCount, scanCount, err := w.backfillIndexInTxn(handleRange)
+		if err == nil {
+			if !w.isWorkerRunnable() {
+				err = errReorgWorkerNotRunnable
+			} else {
+				// Because reorgIndexTask may run a long time,
+				// we should check whether this ddl job is still runnable.
+				err = w.d.isReorgRunnable()
```

Comment:
`d.reorgCtx.notifyCancelReorgJob` is notified only once, it should not be received by one of the workers.
Instead, we can use a variable shared by all workers to notify them the job is canceled.

Reply:
Actually, not only notified once, it will notify reorgJob to exit when we got an errWaitReorgTimeout from runReorgJob. It workers are still working, it will notify continuously.

Reply:
This notification will be sent every time when we run `runDDLJob` before the reorg job converts to rollback job.